### PR TITLE
Flatten ReleaseableBytesReference Object Trees

### PR DESCRIPTION
**Pull request by original-brownbear**
When slicing a releasable bytes reference we would create a new counter
every time and pass the original reference chain to the new slice on every
slice invocation. This would lead to extremely deep reference chains and
needlessly uses a dedicated counter for every slice when all the slices
eventually just refer to the same underlying bytes and `Releasable`.
This commit tracks the ref count wrapper with its releasable in a separate
object that can be passed around on every slicing, making the slices' tree
as flat as the original releasable bytes reference.

Also, we were needlessly creating a redundant releasable bytes reference from
a releasable bytes-stream-output that we never actually used for releasing (all code
that uses it just releases the stream itself instead).

See e.g. the kind reference created by `InboundPipeline` when reading messages and repeatedly slicing:

Screenshot 2020-05-23 at 21 17 52

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/bytes/ReleasableBytesReference.java [hunk 1]
```diff
@@ -34,41 +33,44 @@
...
-        incRef();
+        refCounted.incRef();
         return this;
     }
 
     public ReleasableBytesReference retainedSlice(int from, int length) {
         BytesReference slice = delegate.slice(from, length);
-        incRef();
-        return new ReleasableBytesReference(slice, this);
+        refCounted.incRef();
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/common/bytes/ReleasableBytesReference.java:67]: I would prefer to move this call to the private `ReleasableBytesReference` constructor.

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/default-distro  (Jenkins network issue)

**Comment by original-brownbear** [comment]
Thanks Yannick!

**Pull request merged**